src/App.tsx

```diff
@@ -1,250 +1,219 @@
+            }
+            catch (err) {
+                if (fileHasNotChanged && err instanceof Error) {
+                    setError(err);
+                    setMarkerParsingError(true);
+                }
+            }
+        })();
+        /* Clean-up function called if there is a new file state while parsing is still in progress.
+         * Effectively abandons stale results. */
+        return () => {fileHasNotChanged = false};
+    }, [markerFile]);
+
+    /* Parse provided forceFile into forceFileData */
+    useEffect(() => {
+        /* Only async parse call using the most recent file state should set forceFileData */
+        let fileHasNotChanged = true;
+        /* Call parseForceFileData inside anonymous wrapper function to handle returned Promise or Error */
+        if (forceFile) (async () => {
+            try {
+                const data = await parseForceFileData(forceFile);
+                if (fileHasNotChanged) {
+                    setError(null);
+                    setForceParsingError(false);
+                    setForceFileData(data);
+                }
+            }
+            catch (err) {
+                if (fileHasNotChanged && err instanceof Error) {
+                    setError(err);
+                    setForceParsingError(true);
+                }
+            }
+        })();
+        /* Clean-up function called if there is a new file state while parsing is still in progress.
+         * Effectively abandons stale results. */
+        return () => {fileHasNotChanged = false};
+    }, [forceFile]);
+
+    // --------------------------------------- Animation & Animation Control Data --------------------------------------
+
+    /* Start, current, and end frames from data */
+    const frameStart = 0;
+    const [frame, setFrame] = useState(0);
+    const [frameEnd, setEnd] = useState(0);
+
+    /* User-selected start and end frames (for cropping) */
+    const [frameCropStart, setCropStart] = useState(0);
+    const [frameCropEnd, setCropEnd] = useState(0);
+
+    /* Current frame reference for use in dependency arrays where we don't want to trigger on every frame change */
+    const frameRef = useStateRef(frame);
+
+    /* Flags for playback */
+    const [playing, setPlaying] = useState(false);
+    const [loopPlayback, setLoopPlayback] = useState(true);
+
+    /* Handle on most recent animation loop, so it may be cancelled */
+    const animationRef = useRef<number>();
+
+    /* The time for one marker/force frame in ms */
+    const timeStepMillis = useMemo(() => {
+        if (markerFileData.frames.length<2) return null;
+        return markerFileData.frames[1].time * MILLIS_PER_SEC; //convert from secs to millis
+    }, [markerFileData]);
+
+    /* Time of previous animation loop iteration (most recent render).
+     * For finding elapsed time between loops to determine which marker/force frame to render next. */
+    const lastRepaintTimeRef = useRef<DOMHighResTimeStamp|null>(null);
```

Comment:
not really a repaint since this loop is separate from the one for rendering

Reply:
It seemed to me that the documentation for requestAnimationFrame (which we are calling our animationLoop with) indicates the animation loop runs before each repaint (hence the time between animation loops would be the time since the last repaint):
```
The window.requestAnimationFrame() method tells the browser that you wish to perform an animation
and requests that the browser calls a specified function to update an animation before the next repaint.
The method takes a callback as an argument to be invoked before the repaint...
```

Maybe I'm missing something here that you have a better understanding of, so feel free to jump in and make adjustments.